aesd-char-driver/main.c

```diff
@@ -0,0 +1,202 @@
+    final_count += count;
+
+    /*Write to the circular command buffer*/
+    struct aesd_buffer_entry *new_entry = kmalloc(sizeof(struct aesd_buffer_entry), GFP_KERNEL);
+    if (new_entry == NULL) {
+        retval = -ENOMEM;
+    }
+    else {
+        new_entry->size = final_count;
+        new_entry->buffptr = kmalloc(final_count, GFP_KERNEL);
+        memcpy(new_entry->buffptr, final_buffptr, final_count);
+
+        /*Lock before writing*/
+        mutex_lock(&aesd_device.write_lock);
+
+        /*Write to the buffer*/
+        const char* overwritten_buffptr = aesd_circular_buffer_add_entry(aesd_device.command_buffer, new_entry);
+
+        /*Unlock after writing*/
+        mutex_unlock(&aesd_device.write_lock);
+
+        /*If overwritten, free the overwritten entry*/
+        if(overwritten_buffptr != NULL) {
+            kfree(overwritten_buffptr);
+        }
+        retval = final_count;
+        final_count = 0;
+    }
+    return retval;
+}
+struct file_operations aesd_fops = {
+    .owner =    THIS_MODULE,
+    .read =     aesd_read,
+    .write =    aesd_write,
+    .open =     aesd_open,
+    .release =  aesd_release,
+};
+
+static int aesd_setup_cdev(struct aesd_dev *dev)
+{
+    int err, devno = MKDEV(aesd_major, aesd_minor);
+
+    cdev_init(&dev->cdev, &aesd_fops);
+    dev->cdev.owner = THIS_MODULE;
+    dev->cdev.ops = &aesd_fops;
+    err = cdev_add (&dev->cdev, devno, 1);
+    if (err) {
+        printk(KERN_ERR "Error %d adding aesd cdev", err);
+    }
+    return err;
+}
+
+int aesd_init_module(void)
+{
+    dev_t dev = 0;
+    int result;
+    result = alloc_chrdev_region(&dev, aesd_minor, 1, "aesdchar");
+    aesd_major = MAJOR(dev);
+    if (result < 0) {
+        printk(KERN_WARNING "Can't get major %d\n", aesd_major);
+        return result;
+    }
+    memset(&aesd_device,0,sizeof(struct aesd_dev));
+
+    PDEBUG("Initializing AESD char driver");
+
+    /*Allocate the circular buffer*/
+    struct aesd_circular_buffer* buffer = kmalloc(sizeof(struct aesd_circular_buffer), GFP_KERNEL);
+    aesd_device.command_buffer = buffer;
+    mutex_init(&aesd_device.write_lock); //initialize the mutex
+    aesd_circular_buffer_init(aesd_device.command_buffer); //initialize the buffer
+    aesd_device.command_buffer->in_offs = 0;
+    aesd_device.command_buffer->out_offs = 0;
+    aesd_device.command_buffer->full = false;
+    final_buffptr = kmalloc(4, GFP_KERNEL); //initialize the final_buffptr
+
+    result = aesd_setup_cdev(&aesd_device);
+
+    if( result ) {
+        unregister_chrdev_region(dev, 1);
+    }
+    return result;
+}
+
+void aesd_cleanup_module(void)
+{
+    dev_t devno = MKDEV(aesd_major, aesd_minor);
+
+    cdev_del(&aesd_device.cdev);
+
+    /*Free each entry and its buffptr string in the circular buffer*/
+    struct aesd_buffer_entry* entry;
+    uint8_t index;
+    AESD_CIRCULAR_BUFFER_FOREACH(entry, aesd_device.command_buffer, index) {
+        kfree(entry->buffptr);
+        kfree(entry);
```

Comment:
@ashwinhr186 I don't think you need/want a kfree of the `entry` value, since that isn't dynamically allocated within the circular buffer structure at https://github.com/cu-ecen-aeld/assignments-3-and-later-ashwinhr186/blob/a5faa568666e1b2063b7bbaab310912ad73fcb7e/aesd-char-driver/aesd-circular-buffer.h#L37